Rewritten TS model with segmented laser geometry

Mateasek opened the pull request:
I decided to do new PR, because at the end there were many changes:

- laser geometry is now segmented cylinder as suggested by @mattngc. This propagated into handling of information between plasma, laser node, laser model and scattering model. This was neccessary to simplify handling of notifications, deletions and etc.
  -  laser model and spectrum hold no plasma and laser reference. Their methods return requested properties (e.g. laser power) in laser node frame.
  - Laser material now obtains all the information from plasma, laser spectrum and laser model needed by the scattering model.
  - Scattering model works in the frame of reference of the laser node. Method for scattered spectrum calcullation must now get all the neccessary plasma and laser variables (e.g. Te, ne, directions, etc.)
- laser models were rewritten and new added (uniform, bivariate and trivariate normal spatial distributions, Gaussian beam model). Lasers are now described by pulse energy and pulse temporal lengh, this is transformed to spatial power densities using given distributions.
- laser spectrum power spectral density was corrected.

Unittests checking the following were added:

- laser node
  - initialisation
  - propagation of notifications between laser model, laser spectrum, scattering model, laser geometry and laser material
  - position of the laser segments compared to the expected one
- laser spectrum
  - initialisation and changes
  - normalisation of models
- laser models
  - initialisation
  - polarisation direction
  - correct values of power for spatial points
  - integration of laser power over pulse spatial extent is compared to expected pulse energy
- scattering model
  - scaterred spectrum traced by a ray is compared to expected values calculated by semi-analytical approach.

I hope that this time I minimised PEP8 incompatibilitity. If this state of the model is acceptable and won't be changed much, I will move to writting docstrings, documentation and examples:

- [x] Docstrings
- [x] Documentation
- [x] Examples

Mateasek committed:
Multiple changes:

Segmented cylinder as laser geometry
rewritten laser model to correct power units
new laser model
laser material obtains all properties needed for scattering calculation
laser_spectrum and laser_model independent of plasma or laser
added tests for laser

Mateasek committed:
Multiple changes:

Segmented cylinder as laser geometry
rewritten laser model to correct power units
new laser model
laser material obtains all properties needed for scattering calculation
laser_spectrum and laser_model independent of plasma or laser
added tests for laser

Mateasek committed:
remove empty method

Mateasek committed:
remove empty method

CnlPepper commented:
I'm going to propose we merge this into a feature branch first, so we can make a few tweaks before the core merge.

CnlPepper commented:
This PR is looking a lot better and a lot cleaner. I see there are potentially some areas of improvement in the notification/change handling. I should be able to address that after merge if necessary, provided I'm sure of the original intentions. There are a few things I'd tweak. The material does seem a little over complicated. Some of the properties on the material should really be on the laser etc...

Mateasek committed:
Merge branch 'development' into TS/development

Mateasek committed:
Merge branch 'development' into TS/development

Mateasek committed:
Change: material and references

Laser material simplified. Most of the calculations moved to the scattering model. Caches of plasma and laser transforms
ScatteringModel got reference to the plasma and laser_model to provide values for the scattering calculations
ScatteringModel changed to LaserEmissionModel
Model manager handling emission models added to laser
Propagation of changes and notifications changed to be more like the beam model.

Mateasek committed:
Change: material and references

Laser material simplified. Most of the calculations moved to the scattering model. Caches of plasma and laser transforms
ScatteringModel got reference to the plasma and laser_model to provide values for the scattering calculations
ScatteringModel changed to LaserEmissionModel
Model manager handling emission models added to laser
Propagation of changes and notifications changed to be more like the beam model.

Mateasek committed:
Optimise code, remove redundant imports and weak refs

Mateasek committed:
Optimise code, remove redundant imports and weak refs

Mateasek committed:
Merge branch 'TS/development' into feature/ThomsonScattering

Mateasek commented:
I have simplified the material to be closer to the beam material as suggested by @CnlPepper. The code was moved to the scattering emission calculation. I have inspired myself in the beam model and simplified the notifications.

What should be the next step? Should I remove the draft flag so it is merged to the feature branch?

CnlPepper commented:
I think we should get it in a feature branch and do any final iterations there. I'm going to do a quick check over now.

CnlPepper commented:
I've added some comments to the existing review.

CnlPepper commented:
Hey @Mateasek I see you marked this as ready for review, however there are some outstanding queries above.

Mateasek committed:
Merge branch 'feature/ThomsonScattering' of https://github.com/Mateasek/cherab_core into feature/ThomsonScattering

Mateasek committed:
Fix scattering constants and laser units

Mateasek committed:
fix imports, rename constants

Mateasek committed:
fix more imports

Mateasek committed:
rename LaserModel to LaserProfile

move files model* to profile*

Mateasek committed:
change class names to reflect unit changes

Mateasek committed:
Fix naming

Mateasek committed:
adjust tests to latest changes

Mateasek committed:
rename LaserEmissionModel to LaserModel

move scattering.p* to model.p*
add tests/.__init__.py
adjust test_laser.py unittests to match latest changes

Mateasek committed:
Add missing line

Mateasek committed:
remove internal use of notifier.notify()

Mateasek commented on cherab/core/laser/node.pyx:
```diff
@@ -0,0 +1,233 @@
+        self.notifier.add(self._configure_geometry)
+
+        #setup model manager
+        self._models = ModelManager()
```
I added self._configure_geometry to manager notifier

Mateasek replied:
I moved the geometry from the Laser node to the LaserProfile class, this will allow to do more effective geometries for the laser shapes in the future.

Mateasek reviewed (commented):
Thanks @CnlPepper  for the review. I tried to address all the comments and did changes to follow the suggestions. There is one I am not sure is correct and that is removing self._configure_geometry from the notifier.

Mateasek committed:
add _configure_geometry to laser.notifier

Mateasek committed:
Merge branch 'development' into feature/ThomsonScattering

Mateasek committed:
remove _configure_geometry from laser notifier

Add notifier notification to the end of the _configure_geometry to
notify about changes

Mateasek committed:
Merge branch 'development' into feature/ThomsonScattering

Mateasek committed:
fix constant problems

Fix bug in calculation of absolute values.

LaserProfiles now provide energy density
Fixed bug in Selden approximation. Corrected normalisation with laser
wavelength and also selden approximation already returns density values,
no need to normalise by delta lamda.

Mateasek committed:
Move geometric properties to laser profile

It is better for the laser geometry to be bound to the laser profiles
rather than laser node. It will allow more efficient shaping of the
laser primitives.

Mateasek committed:
Fix tests, add laser profile tests

Mateasek commented:
I finally managed to complete the code. I fixed some problems with constants and moved the geometry generation from the Laser class to the LaserProfile class, this should allow using more efficient geometries for laser profiles in the future. Whenever @CnlPepper has time, I think this is ready for another go.

Mateasek committed:
Merge branch 'development' of https://github.com/cherab/core into feature/ThomsonScattering

Mateasek committed:
Merge commit '68bc1887656f2445891f499cea62fb3b411127ee' into feature/ThomsonScattering

Mateasek committed:
add dosctring to the LaserSpectrum class

Mateasek committed:
Add dosctring to missing properties.

Mateasek committed:
Add dosctrings to laser spectra

Mateasek committed:
Add docstrings to the LaserProfile

Mateasek committed:
Merge branch 'development' of https://github.com/cherab/core into feature/ThomsonScattering

Mateasek committed:
Add docstring to LaserSpectrum base

Mateasek committed:
add missing property docstrings

Mateasek committed:
Add docstring to LaserSpectrum classes

Mateasek committed:
Add doctring to math_functions used in laser.

Mateasek committed:
Add docstring to LaserProfile base

Mateasek committed:
Add docstring to laser profiles

Mateasek committed:
Add angular dependence in polarisation plane

Intensity of scattered light follows cos2 distribution in the plane of
polarisation and perpendicular to laser propagation.

Mateasek committed:
Add docstring to laser models

fix naming and header of methods

Mateasek committed:
Fix error in the polarisation angle

Mateasek committed:
Add Generomak Core plasma

a set of functions which provide core plasma for Generomak were added.
The plasma population profiles follow either a double quadratic profil,
which is used for temperatures and some densities (H1+, C6+) and
exponentially decaying profiles for densities. The Profiles are mapped
into 3D using the generomak equilibrium and assuming constant values
along magnetic surfaces.

Mateasek committed:
Add demo with Generomak core profiles plots.

Mateasek committed:
Merge branch 'development' into feature/ThomsonScattering

Mateasek committed:
Merge branch 'development' into generomak/core_plasma

Mateasek committed:
Merge branch 'development' into feature/ThomsonScattering

Mateasek committed:
Add docstrings to Laser node.

Mateasek committed:
Remove empty line

Mateasek committed:
Move laserspectrum files to models folder

Mateasek committed:
Move laserspectrum base files

Mateasek committed:
Change name to photon_spectral_density

photons attributes and property renamed to photon_spectral_density
which is more descriptive of the value meaning

Mateasek committed:
Rename attributes, change calculation of PSD

rename photon related attributes
Change the way how power spectral density of the laser spectrum
is calculated. The current approach should be more precise and allows
users to define a more precise formula when overriding the
_get_bin_power_spectral_density method. This was done for the
gaussian spectrum.

Mateasek committed:
Add __init__ files to cherab/core/model/laser

Mateasek committed:
Move profile files to proper locations

Mateasek committed:
Move math_function files to proper location

Mateasek committed:
Fix imports

Mateasek committed:
Improve docstrings

Mateasek committed:
Applying PR requested changes

A lot of typo removed from docstrings
Chaged plasma geometry z transfer to the correct value
removed font changes from plots

Mateasek committed:
Improve docstrings, remove typos

Mateasek committed:
Add laser profile demos

shows the basic versions of laser profiles with plots

Mateasek committed:
Add Laser spectra demo with plots

Mateasek committed:
Separate LaserModel baseclass

Mateasek committed:
Move SeldenMatoba to proper location

Mateasek committed:
Move laserspectrum tests

Mateasek committed:
Move test_laserspectrum, to the correct place...

Mateasek committed:
Improve tests and fix LaserSpectrum bug

LaserSpectrum min_wavelength should be lower than max_wavelength,
not lower or equal
Fixed, completed and improved LaserSpectrum tests

Mateasek committed:
fix pep8

Mateasek committed:
Futher improve LaserSpectrum base tests

Mateasek committed:
Add laserspectra unittests

Test if ConstantSpectrum and GaussianSpectrum are normalised

Mateasek committed:
Move LaserProfile unittests

Mateasek committed:
Rename file to match class names

Mateasek committed:
Fix method name

Mateasek committed:
Fix bugs and rename

rename CONST_TS to RATE_TS to make the name more explanatory.
the meaning of the value is reaction rate coefficient.
Fix the incorrect orientation of dipole radiation. If the polarisation
is the direction of the electric field, then the dependence is sin ** 2

Mateasek committed:
Add unittest to SeldenMatobaThomsonSpectrum

Mateasek committed:
Remove redundant imports

Mateasek committed:
Use integrator from  passed arguments

Mateasek committed:
Improve docstring

Clarify what generate_geometry method is expected to return

Mateasek committed:
Add segmented cylinder tests

Mateasek committed:
Adjust Laser node tests

Tests of models,profiles and spectra moved to separate files and proper
locations. Adjusted Laser node tests.

Mateasek committed:
Improve checks and docstring

Corrected docstring
Adjusted checks in models.setter to be more pythonic.

Mateasek committed:
Add __init__.py imports

Mateasek committed:
Remove redundant __init__ files

Mateasek committed:
PEP8

Mateasek committed:
Add SeldenMatoba Thomson scattering demo

Mateasek committed:
PEP8

Mateasek committed:
Merge branch 'generomak/core_plasma' into feature/ThomsonScattering

Mateasek committed:
Add imports of core plasma to __init__.py

Mateasek committed:
Merge branch 'generomak/core_plasma' into feature/ThomsonScattering

Mateasek committed:
Add Thomson scattering demo

Mateasek committed:
Add copyright header

Mateasek commented on cherab/core/laser/scattering.pyx:
```diff
@@ -0,0 +1,168 @@
+
+cdef class LaserEmissionModel:
+
+    def __init__(self, Laser laser):
```
Changed the name of the class to LaserModel, changed the __init__ to follow the beam way.

Mateasek committed:
Merge branch 'development' into feature/ThomsonScattering

Mateasek commented:
If someone updates or changes the target branch, I this is ready for review, whenever @CnlPepper , @jacklovell @vsnever find some time.

Mateasek committed:
remove forgotten todos in docstring

Mateasek committed:
Add documentation to Laser and others

Mateasek commented on cherab/core/laser/node.pyx:
```diff
@@ -0,0 +1,167 @@
+                self._geometry.append(cylinder)
+
+        # length of the last segment is laser_length - n * 2 radius to avoid infinitesimal reminder segments
+        segment_length_last = self.length - (n_segments - 1) * segment_length
```
Changed, geometry generating method was moved to the LaserProfile class. The reason is LaserProfile holds information about the laser energy density and can generate optimal primitives to create an envelope.

Mateasek commented:
I resolved all the comments. Most of them I tried to answer or were obsolete already. Could someone look at it? @jacklovell , @vsnever @CnlPepper @mattngc

vsnever reviewed (changes requested):
Hi @Mateasek,

This is a very important new feature. As far as I know, the Thomson scattering synthetic diagnostic is in great demand at ITER. Thanks a lot for this!

The comments are many but all the issues are minor.
Also, could you please update the changelog?

Mateasek committed:
Change imports

Mateasek commented on cherab/core/laser/__init__.py:
```diff
@@ -0,0 +1 @@
+from .node import Laser
```
requested changes applied

Mateasek committed:
Fix typo

Mateasek committed:
Remove photon related quantities from LaserSpectrum

Mateasek commented on cherab/core/laser/laserspectrum.pyx:
```diff
@@ -0,0 +1,193 @@
+    def power_spectral_density(self):
+        return self._power_spectral_density
+
+    @property
+    def photon_spectral_density(self):
+        return self._photon_spectral_density
```
This is a left-over, thanks for pointing it out. Removed.

Mateasek committed:
Remove laser reference from LaserProfile

There is currently no need to keep laser reference n LaserProfile. This
was a leftover from version, where some laser properties were in Laser

Mateasek commented on cherab/core/laser/profile.pyx:
```diff
@@ -0,0 +1,181 @@
+
+        return self._energy_density3d.evaluate(x, y, z)
+
+    @property
+    def laser(self):
+        return self._laser
+    
+    @laser.setter
+    def laser(self, value):
+        if not isinstance(value, Laser):
+            raise TypeError("Value has to instance of Laser class.")
+        
+        if self._laser is not None:
+            self.notifier.remove(self._laser._configure)
+        
+        self._laser = value
+        self.notifier.add(self._laser._configure)
+        self._change()
+        
+        self.notifier.notify()
```
Yes, the reference is again a leftover from previous version. Thanks for pointing this out, now the reference is completely useless. I'll remove it.

Mateasek committed:
Remove reversed logic

Mateasek committed:
Remove more inverse logic

Mateasek commented on cherab/core/model/laser/model.pyx:
```diff
@@ -0,0 +1,222 @@
+        :return: Spectrum
+        """
+        # check for nonzero laser power, ne, te, wavelength
+        if not ne > 0 or not te > 0 or not laser_energy_density > 0:
```
removed

Mateasek commented on cherab/core/model/laser/math_functions.pyx:
```diff
@@ -0,0 +1,300 @@
+
+    @stddev_waist.setter
+    def stddev_waist(self, double value):
+        if not value > 0:
```
removed

Mateasek commented on cherab/core/model/laser/math_functions.pyx:
```diff
@@ -0,0 +1,300 @@
+
+    @wavelength.setter
+    def wavelength(self, double value):
+        if not value > 0:
```
removed

Mateasek commented on cherab/core/laser/laserspectrum.pyx:
```diff
@@ -0,0 +1,193 @@
+
+        if not min_wavelength > 0:
+            raise ValueError("min_wavelength has to be larger than 0, but {} passed.".format(min_wavelength))
+        if not max_wavelength > 0:
```
removed

Mateasek commented on cherab/core/laser/laserspectrum.pyx:
```diff
@@ -0,0 +1,193 @@
+
+    def _check_wavelength_validity(self, min_wavelength, max_wavelength):
+
+        if not min_wavelength > 0:
```
removed

Mateasek commented on cherab/core/laser/laserspectrum.pyx:
```diff
@@ -0,0 +1,193 @@
+
+    @bins.setter
+    def bins(self, int value):
+        if not value > 0:
```
removed

Mateasek commented on cherab/core/model/laser/model.pyx:
```diff
@@ -0,0 +1,222 @@
+        # check for nonzero laser power, ne, te, wavelength
+        if not ne > 0 or not te > 0 or not laser_energy_density > 0:
+            return spectrum
+        if not laser_wavelength >= 0:
```
removed

Mateasek commented on cherab/core/laser/node.pyx:
```diff
@@ -0,0 +1,249 @@
+        if self._laser_profile is None:
+            return
+
+        # update the plasma reference in models
+        for model in list(self._models):
+            model.plasma = self._plasma
```
removed

Mateasek commented on cherab/core/model/laser/math_functions.pxd:
```diff
@@ -0,0 +1,48 @@
+cdef class ConstantAxisymmetricGaussian3D(Function3D):
+
+    cdef:
+        double _stddev, _recip_negative_2stddev2, _recip_2pistddev2
```
fixed the naming in pxd file

Mateasek committed:
Add minor codereview suggestions and fixes

Mateasek commented on cherab/core/model/laser/laserspectrum.pyx:
```diff
@@ -0,0 +1,133 @@
+
+    @mean.setter
+    def mean(self, double value):
+        if not value > 0:
```
removed

Mateasek committed:
Apply multiple code review suggestions

minor ones

Mateasek committed:
Replace leading _ in attributes

In LaserSpectrum class

Mateasek commented on cherab/core/laser/laserspectrum.pxd:
```diff
@@ -0,0 +1,41 @@
+        double _min_wavelength, _max_wavelength, _delta_wavelength
+        readonly int _bins
+        ndarray _power, _power_spectral_density, _photon_spectral_density, _wavelengths  # power_spectral_density [w/nm]
+        double[::1] _power_mv, _power_spectral_density_mv, _photon_spectral_density_mv, _wavelengths_mv
```
I removed leading _ in case of the memoryviews, which are to be used outside the class. I kept the name _bins, but added accessors: get_spectral_bins, get_min_wavelength, get_max_wavelength and get_delta_wavelength to expose the values outside of the class.

Mateasek committed:
Change accesing private attribute

from laser._plasma to laser.plasma

Mateasek commented on cherab/core/laser/material.pyx:
```diff
@@ -0,0 +1,66 @@
+        super().__init__(integrator)
+
+        self._laser_segment_to_laser_node = laser_segment.to(laser)
+        self._laser_to_plasma = laser_segment.to(laser._plasma)
```
changed to laser.plasma

Mateasek commented on cherab/core/model/laser/model.pyx:
```diff
@@ -0,0 +1,222 @@
+        polarisation_vector = self._laser_profile.get_polarization(laser_x, laser_y, laser_z)
+        angle_polarization = observation_laser.angle(polarisation_vector) # scattering direction is the opposite to obervation direction
+
+        laser_wavelength_mv = self._laser_spectrum._wavelengths_mv
+        laser_spectrum_power_mv = self._laser_spectrum._power_mv  # power in spectral bins (PSD * delta wavelength)
+        bins = self._laser_spectrum._bins
```
solved by removing leading _ in memoryviews of `laser_spectrum`, `self._laser_spectrum._bins` replaced with `self._laser.get_spectral_bins()`

Mateasek committed:
Changed accessing properties to attributes

Mateasek committed:
Change notifications and geometry rebuilding

Rebuilding segment primitives was separated from assigning materials to
segment primitives. Rebuilding primitives is now called only in needed
cases.

Added configure_geometry to laser_profile notifier
Fixed missing addition of _plasma_changed to plasma notifier

Changing integrator changes only integrators within materials

Mateasek committed:
Add tests to check if integrators are changed

Mateasek commented on cherab/core/laser/node.pyx:
```diff
@@ -0,0 +1,249 @@
+    def plasma(self, Plasma value not None):
+
+        #unregister from old plasma notifier
+        if self._plasma is not None:
+            self._plasma.notifier.remove(self._plasma_changed)
+
+        self._plasma = value
```
Thanks for spotting this! This could definitely lead to serious problems. `Plasma` has to notify `Laser`, so it rebuilds the materials, because  `LaserMaterial` caches transforms to plasma.  I separated the method `_configure_geometry` into `_build_geometry`, which now only builds the primitives and into `_configure_materials`, which updates the materials. The `_configure_geometry` was replaced by `configure_geometry`

Mateasek commented on cherab/core/laser/node.pyx:
```diff
@@ -0,0 +1,249 @@
+            self._plasma.notifier.remove(self._plasma_changed)
+
+        self._plasma = value
+        self._configure()
```
changed to updating materials only: `self._configure_materials` is now called.

Mateasek commented on cherab/core/laser/node.pyx:
```diff
@@ -0,0 +1,249 @@
+    def importance(self, double value):
+        
+        self._importance = value
+        self._configure()
```
changed to updating materials only: `self._configure_materials` is now called.

Mateasek commented on cherab/core/laser/node.pyx:
```diff
@@ -0,0 +1,249 @@
+    @laser_spectrum.setter
+    def laser_spectrum(self, LaserSpectrum value):
+        self._laser_spectrum = value
+        self._configure()
```
changed to updating materials only: `self._configure_materials` is now called.

Mateasek commented on cherab/core/laser/node.pyx:
```diff
@@ -0,0 +1,249 @@
+            raise ValueError("The plasma, laser_profile and laser_spectrum must be set before before specifying any models.")
+
+        self._models.set(value)
+        self._configure()
```
changed to updating materials only: `self._configure_materials` is now called.

Mateasek commented on cherab/core/laser/node.pyx:
```diff
@@ -0,0 +1,249 @@
+    @integrator.setter
+    def integrator(self, VolumeIntegrator value):
+        self._integrator = value
+        self._configure()
```
suggestion applied

Mateasek commented on cherab/core/laser/node.pyx:
```diff
@@ -0,0 +1,249 @@
+    def get_geometry(self):
+        return self._geometry
+
+    def _plasma_changed(self):
+        """React to change of plasma and propagate the information."""
+        self._configure()
```
I kept the `_plasma_changed` which calls `self._configure_materials`

Mateasek commented on cherab/core/laser/node.pyx:
```diff
@@ -0,0 +1,249 @@
+        """React to change of plasma and propagate the information."""
+        self._configure()
+
+    def _modified(self):
+        self._configure()
```
`_modified` is defined [here](https://github.com/raysect/source/blob/3c1f2f2f1cce9c2a2d5ca2952a70af3287d53dbf/raysect/core/scenegraph/_nodebase.pyx#L174). Because there are transforms cached in material, it should be rebuilt when the `Laser` transform changes I think

vsnever replied:
I see, thanks.

Mateasek committed:
Fix wrong docstring

Mateasek commented on cherab/core/model/laser/profile.pyx:
```diff
@@ -0,0 +1,776 @@
+        the laser_energy is spread along the z axis using the velocity
+        of light SPEED_OF_LIGHT and the temporal duration of the pulse:
+        length = SPEED_OF_LIGTH * pulse_length. Combining the normal distribution with the normalisation
+         pulse_energy / length gives the units [J / m ** 3].
```
fixed the doctring, thanks

Mateasek commented:
Thank you very much @vsnever for doing this very thorough review and for suggesting the improvements and spotting problems. In case of the minor suggestions I applied the suggestions and marked them as resolved. There are few regarding the notifications and rebuilding of the geometry and materials, which I left opened because the applied changes should be checked again. I would like to ask you to see if it is what you had in mind, thanks.

Mateasek committed:
Update changelog

vsnever commented on cherab/core/laser/profile.pxd:
```diff
@@ -0,0 +1,42 @@
+    cdef:
+        VectorFunction3D _polarization3d, _pointing3d
+        Function3D _energy_density3d
+        Laser _laser
```
The `_laser` is a leftover attribute. `LaserProfile` no longer holds a reference to the laser.

vsnever commented on cherab/core/model/laser/model.pxd:
```diff
@@ -0,0 +1,40 @@
+cdef class SeldenMatobaThomsonSpectrum(LaserModel):
+    
+    cdef:
+        double re, _CONST_ALPHA, _RATE_TS, _RECIP_M_PI
```
Looks like `re` is a leftover attribute.

vsnever reviewed (commented):
Thanks for fixing all the issues. Now everything looks fine, except for a couple of leftover attributes.

Mateasek committed:
Remove redundant attributes

vsnever reviewed (approved):
Thanks @Mateasek, I'm happy with this PR.

Mateasek committed:
Merge branch 'development' into feature/ThomsonScattering

Mateasek committed:
Remove empty line

jacklovell committed:
Merge branch 'development' into feature/ThomsonScattering
